### DISCUSSION TITLE
Is this proper way to round a floating point value to an integer?

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [ispc/ispc#2776] [Q&A]
My concern is that the `round` function still returns a floating point value, so if the value can not be represented as a whole integer, then casting to integer will truncate the decimal part which leads to incorrect behavior (?). this compiles to `vroundps` and `vcvttps2dq`.

```c++
inline varying int32 RoundToInt(varying float x)
{
    return (varying int32)round(x);
}
```

so im not sure if it's safe or not? the other way would be

```c++
inline varying int32 RoundToInt(varying float x)
{
    varying float Half = floatbits((intbits(x) & 0x80000000) | intbits(0.5f));
    return (varying int32)(x+ Half); // return (int32)(x + (x>0 ? 0.5 : -0.5))
}
```

This is at least safe but a little slower. I would love to know if round function does the job or I need to implement my own rounding function like this.

**Comment by nurmukhametov** [comment]
What do you mean with incorrect behavior and safe? There are integer values that are representable in `float` type but unrepsentable in `int` type. Convertation for these values is incorrect behavior (e.g., in C standard it is undefined behavior).

I do not understand what problem your code solves. I would do just casting from `float` to `int`. 
```ispc
varying int32 RoundToInt(varying float x)
{
    return x;
}
```
It generates the only `vcvttps2dq` that truncates (rounds towards zero). Why is it not enough in your case?

**Reply by MkazemAkhgary** [reply]
I actually want a "round to nearest integer" function. so for example `round(1.9) == 2`. now I see that my question was wrong and I was being stupid and overthinking.

what I meant was some thing like this, Lets say `round(13.3)` becomes `12.999` and then casting `12.999` to integer would give me `12` instead of `13`. although this is a wrong example and 13 can be represented by a float.

but it seems that round function would return nearest integer representable by a float and what ever that would be, it would get casted to the integer. now I see that my second RoundToInt does not solve anything 😄 

I checked and 32 bit floats can represent up to 16777216 without any gaps in integers. my numbers are much lower than this. so there should be no problem in my case.
